File: python/bridge/torch_worker.py
"""Long-lived torch worker that hosts the GPU/torch bridge CLIs so their models stay warm.

This is staged-rollout **step 4** of ``docs/cards/editor-resource-model.md``
("torch long-lived Python worker: Rust spawns and keeps it alive; replaces
per-call subprocess + model reload for realesrgan / sd_inpaint").

The Rust host (``studio/torch_worker.rs``) spawns exactly one of these and keeps
it alive for the life of the desktop process, then talks to it with
newline-delimited JSON: one request object per line on ``stdin``, one response
object per line on ``stdout``. Because the process is long-lived, the torch
backends' process-global warm caches (``sr_backends.realesrgan._WARM_UPSAMPLERS``,
``inpaint_backends.sd_inpaint._WARM_PIPELINES``) survive across requests, so the
~64 MB Real-ESRGAN weight / multi-GB SD-inpaint pipeline is built once per
``(weight, device, precision)`` instead of reloaded on every run — the dominant
latency the old per-call subprocess paid.

Protocol (one JSON object per line, UTF-8):

    request : {"id": <any>, "cmd": "ping"|"image_enhance"|"detail_repaint"|"shutdown",
               "argv": [str, ...]}
    response: {"id": <same>, "ok": bool, "code": int, "stdout": str, "error": str}

``argv`` is the exact argument vector the one-shot CLI would have received, so
the worker is a drop-in host: it invokes the CLI's ``main(argv)`` with
``stdout``/``stderr`` captured and returns the CLI's JSON on ``stdout`` plus its
exit ``code``. All argument parsing, CPU fallback and reporting stay in the CLIs;
the worker adds only warmth (via the backends' caches) and process management.
A request whose ``cmd`` maps to a CLI that fails still yields a well-formed
response (``ok=false`` with the captured ``error``) so one bad run never wedges
the worker — the host keeps sending it more.
"""

from __future__ import annotations

import contextlib
import io
import json
import sys
from typing import Any, Callable

import detail_repaint_cli
import image_enhance_cli

#: ``cmd`` -> the CLI ``main(argv) -> int`` that services it. Both CLIs share the
#: same contract: write a single JSON result line to stdout, a single error line
#: to stderr, and return a POSIX exit code (0 == ok).
_COMMANDS: dict[str, Callable[[list[str]], int]] = {
    "image_enhance": image_enhance_cli.main,
    "detail_repaint": detail_repaint_cli.main,
}


def _run_cli(main: Callable[[list[str]], int], argv: list[str]) -> dict[str, Any]:
    """Invoke a CLI ``main(argv)`` in-process, capturing its stdout/stderr.

    The CLIs write their JSON result to stdout and a single error line to
    stderr; we redirect both so nothing leaks onto the worker's own stdout
    protocol stream. The captured stdout is returned verbatim (the host parses
    it exactly as it parsed the one-shot subprocess output).
    """
    out, err = io.StringIO(), io.StringIO()
    try:
        with contextlib.redirect_stdout(out), contextlib.redirect_stderr(err):
            code = int(main(list(argv)))
    except SystemExit as exc:  # argparse errors call sys.exit; treat as a failed run
        code = int(exc.code) if isinstance(exc.code, int) else 1
    except Exception as exc:  # noqa: BLE001 - a crashed run must not kill the worker
        return {"ok": False, "code": 1, "stdout": out.getvalue(), "error": f"{type(exc).__name__}: {exc}"}
    return {
        "ok": code == 0,
        "code": code,
        "stdout": out.getvalue(),
        "error": err.getvalue().strip(),
    }


def handle_request(request: dict[str, Any]) -> dict[str, Any] | None:
    """Service one decoded request, returning the response (``None`` == shut down).

    Never raises: an unknown ``cmd`` or a crashing CLI yields an ``ok=false``
    response rather than propagating, so the host's read loop stays in lock-step
    (one response per request) and a single bad run cannot wedge the worker.
    """
    req_id = request.get("id")
    cmd = request.get("cmd")

    if cmd == "shutdown":
        return None
    if cmd == "ping":
        return {"id": req_id, "ok": True, "code": 0, "stdout": "", "error": ""}

    main = _COMMANDS.get(cmd) if isinstance(cmd, str) else None
    if main is None:
        return {"id": req_id, "ok": False, "code": 1, "stdout": "", "error": f"unknown cmd {cmd!r}"}

    argv = request.get("argv") or []
    if not isinstance(argv, list) or not all(isinstance(a, str) for a in argv):
        return {"id": req_id, "ok": False, "code": 1, "stdout": "", "error": "argv must be a list of strings"}

    result = _run_cli(main, argv)
    result["id"] = req_id
    return result


def serve(stdin: Any = None, stdout: Any = None) -> int:
    """Run the request/response loop until EOF or a ``shutdown`` request.

    Reads one JSON request per line and writes one JSON response per line,
    flushing after each so the host sees results promptly. A line that is not
    valid JSON gets an ``ok=false`` response (with no ``id``) rather than
    crashing the loop. Returns ``0`` on a clean exit.
    """
    stdin = stdin if stdin is not None else sys.stdin
    stdout = stdout if stdout is not None else sys.stdout

    for line in stdin:
        line = line.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
        except json.JSONDecodeError as exc:
            response: dict[str, Any] | None = {
                "id": None,
                "ok": False,
                "code": 1,
                "stdout": "",
                "error": f"invalid request json: {exc}",
            }
        else:
            response = handle_request(request)
            if response is None:  # shutdown
                break

        stdout.write(json.dumps(response, ensure_ascii=False) + "\n")
        stdout.flush()
    return 0


if __name__ == "__main__":
    raise SystemExit(serve())

File: apps/desktop-tauri/src-tauri/src/studio/torch_worker.rs
//! Long-lived Python worker that keeps the torch models warm (`Compute` lane).
//!
//! The torch bridge CLIs (`image_enhance_cli.py` with `--engine realesrgan`,
//! `detail_repaint_cli.py repaint --engine sd_inpaint`) previously ran as a
//! fresh `python` subprocess per call, and each rebuilt its model from disk —
//! the ~64 MB Real-ESRGAN weight, the multi-GB Stable Diffusion inpaint
//! pipeline — on *every* invocation. This module spawns **one** persistent
//! worker (`python/bridge/torch_worker.py`) and keeps it alive for the life of
//! the process, sending it one request per torch call over piped stdin/stdout.
//! Because the worker process is long-lived, the Python backends' process-global
//! warm caches survive across requests, so those weights load once per
//! `(weight, device, precision)` instead of on every run. This is staged-rollout
//! step 4 of `docs/cards/editor-resource-model.md` ("torch long-lived Python
//! worker: Rust spawns and keeps it alive; replaces per-call subprocess + model
//! reload for realesrgan / sd_inpaint").
//!
//! The worker is guarded by a single `Mutex`, so torch calls are serialised
//! through it — one request is written and its response read before the next
//! begins. That matches the "don't run torch inference concurrently" rule and
//! the GPU `Semaphore(1)` policy from step 2. Like the ONNX warm pool it is a
//! plain process-global `static` (not Tauri managed state) so the handle-free
//! `psd.rs` commands can reach it.
//!
//! Failure is always recoverable: if the worker cannot be spawned, its pipe
//! breaks, or it dies, [`run_cli`] returns `Err` and the caller falls back to
//! the original one-shot subprocess. A broken worker is dropped and respawned
//! on the next call (with one in-call retry across a dead pipe), so a single
//! crash never wedges the feature.

use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdin, ChildStdout, Command, Stdio};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Mutex, OnceLock};

use serde::Deserialize;
use serde_json::json;

/// One request/response the worker understands. `stdout` carries the hosted
/// CLI's JSON (exactly what the one-shot subprocess would have printed) and
/// `code` its POSIX exit status; `ok` is `code == 0`.
#[derive(Debug, Deserialize)]
struct WorkerResponse {
    #[serde(default)]
    ok: bool,
    #[serde(default)]
    code: i64,
    #[serde(default)]
    stdout: String,
    #[serde(default)]
    error: String,
}

/// A running worker: the child process plus its framed stdin/stdout. Pinned to
/// the `(python, dir)` it was spawned for so a project/interpreter change forces
/// a respawn (the worker's cwd is fixed at spawn, and relative image/output
/// paths resolve against it just as the one-shot's `current_dir` did).
struct Worker {
    python: PathBuf,
    dir: PathBuf,
    child: Child,
    stdin: ChildStdin,
    stdout: BufReader<ChildStdout>,
}

impl Worker {
    /// Send one request line and read its single response line. Any I/O or
    /// decode failure here means the worker/pipe is unusable and the caller
    /// should drop and (once) respawn it.
    fn request(&mut self, cmd: &str, argv: &[String]) -> Result<WorkerResponse, String> {
        let id = next_request_id();
        let mut line = json!({ "id": id, "cmd": cmd, "argv": argv }).to_string();
        line.push('\n');
        self.stdin
            .write_all(line.as_bytes())
            .and_then(|_| self.stdin.flush())
            .map_err(|err| format!("torch worker write failed: {err}"))?;

        let mut resp_line = String::new();
        let read = self
            .stdout
            .read_line(&mut resp_line)
            .map_err(|err| format!("torch worker read failed: {err}"))?;
        if read == 0 {
            return Err("torch worker closed its stdout".to_string());
        }
        serde_json::from_str::<WorkerResponse>(resp_line.trim())
            .map_err(|err| format!("torch worker sent invalid json: {err} (raw: {})", resp_line.trim()))
    }
}

impl Drop for Worker {
    fn drop(&mut self) {
        // Best-effort graceful stop: ask the loop to exit, then reap the child
        // so we never leak a zombie when the worker is replaced or the app ends.
        if let Ok(mut line) = serde_json::to_string(&json!({ "cmd": "shutdown" })) {
            line.push('\n');
            let _ = self.stdin.write_all(line.as_bytes());
            let _ = self.stdin.flush();
        }
        let _ = self.child.wait();
    }
}

/// The single warm worker (`None` until first use / after a crash).
static WORKER: OnceLock<Mutex<Option<Worker>>> = OnceLock::new();
/// Monotonic request ids, purely for correlating a response with its request.
static REQUEST_ID: AtomicU64 = AtomicU64::new(1);

fn worker_cell() -> &'static Mutex<Option<Worker>> {
    WORKER.get_or_init(|| Mutex::new(None))
}

fn next_request_id() -> u64 {
    REQUEST_ID.fetch_add(1, Ordering::Relaxed)
}

/// Resolve the worker script under a project's vendored bridge, erroring if the
/// helper is missing (older checkout / bundle) so the caller falls back.
fn worker_script(dir: &Path) -> Result<PathBuf, String> {
    let script = dir.join("python").join("bridge").join("torch_worker.py");
    if !script.is_file() {
        return Err(format!("torch_worker.py not found at {}", script.display()));
    }
    Ok(script)
}

/// Spawn a fresh worker for `(python, dir)` with piped stdin/stdout. The child's
/// cwd is the project dir (matching the one-shot `current_dir`) and its stderr
/// is discarded — errors are reported per-request in the response `error`.
fn spawn(python: &Path, dir: &Path) -> Result<Worker, String> {
    let script = worker_script(dir)?;
    let mut cmd = Command::new(python);
    cmd.arg(&script)
        .current_dir(dir)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null());
    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;
        // CREATE_NO_WINDOW: don't pop a console window for the child.
        cmd.creation_flags(0x0800_0000);
    }

    let mut child = cmd
        .spawn()
        .map_err(|err| format!("failed to launch torch worker {}: {err}", python.display()))?;
    let stdin = child
        .stdin
        .take()
        .ok_or_else(|| "torch worker stdin unavailable".to_string())?;
    let stdout = child
        .stdout
        .take()
        .ok_or_else(|| "torch worker stdout unavailable".to_string())?;
    Ok(Worker {
        python: python.to_path_buf(),
        dir: dir.to_path_buf(),
        child,
        stdin,
        stdout: BufReader::new(stdout),
    })
}

/// Whether the current worker (if any) can serve this `(python, dir)`; a
/// mismatch means a different project/interpreter and forces a respawn.
fn matches(worker: &Option<Worker>, python: &Path, dir: &Path) -> bool {
    worker
        .as_ref()
        .is_some_and(|w| w.python == python && w.dir == dir)
}

/// Run a torch bridge CLI through the warm worker and return its stdout (the
/// hosted CLI's JSON), keeping the models warm across calls.
///
/// `cmd` is the worker command (`"image_enhance"` or `"detail_repaint"`) and
/// `argv` the exact argument vector the one-shot CLI would have received (for
/// `detail_repaint` that includes the `prepare`/`repaint`/`composite`
/// subcommand as `argv[0]`).
///
/// Returns `Err` on two distinct conditions, both of which the caller treats as
/// "use the one-shot fallback": the worker infrastructure is unavailable (spawn
/// failed, pipe broke — retried once with a fresh worker), **or** the hosted CLI
/// itself exited non-zero (surfaced via its captured `error`). Either way the
/// authoritative one-shot subprocess then runs, so behaviour is unchanged when
/// the worker is absent and a genuine CLI error still surfaces to the user.
pub(crate) fn run_cli(python: &Path, dir: &Path, cmd: &str, argv: &[String]) -> Result<String, String> {
    let mut guard = worker_cell()
        .lock()
        .map_err(|_| "torch worker mutex poisoned".to_string())?;

    let mut last_transport_err = String::new();
    // Two attempts: the first may hit a stale/dead pipe from an earlier crash;
    // dropping it and respawning gives one clean retry before we give up.
    for _ in 0..2 {
        if !matches(&guard, python, dir) {
            *guard = None; // drop (and gracefully stop) any mismatched worker
            *guard = Some(spawn(python, dir)?);
        }

        let worker = guard
            .as_mut()
            .expect("worker present after spawn/match check");
        match worker.request(cmd, argv) {
            Ok(resp) if resp.ok => return Ok(resp.stdout),
            Ok(resp) => {
                // The CLI ran but failed (bad args, backend error). This is a
                // real error, not a transport fault, so surface it immediately
                // rather than respawning — a retry would fail identically.
                return Err(if resp.error.is_empty() {
                    format!("torch worker command {cmd} failed (exit {})", resp.code)
                } else {
                    resp.error
                });
            }
            Err(transport) => {
                // Pipe broke / worker died: discard it so the next iteration
                // (or the next call) spawns a fresh one.
                *guard = None;
                last_transport_err = transport;
            }
        }
    }
    Err(last_transport_err)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn worker_script_errors_when_missing() {
        let dir = Path::new("Z:/definitely/missing-project");
        let err = worker_script(dir).unwrap_err();
        assert!(err.contains("torch_worker.py not found"));
    }

    #[test]
    fn matches_is_false_without_a_worker() {
        let none: Option<Worker> = None;
        assert!(!matches(&none, Path::new("python"), Path::new("/proj")));
    }

    #[test]
    fn request_ids_are_monotonic() {
        let a = next_request_id();
        let b = next_request_id();
        assert!(b > a);
    }

    #[test]
    fn response_defaults_are_a_failure() {
        // A worker reply missing every field must decode as not-ok so a garbled
        // response is treated as a failure rather than a silent success.
        let resp: WorkerResponse = serde_json::from_str("{}").unwrap();
        assert!(!resp.ok);
        assert_eq!(resp.code, 0);
        assert!(resp.stdout.is_empty());
    }
}
